src/AcmeClient.php

```diff
@@ -11,79 +11,268 @@
+            'type'             => 'http-01',
+            'keyAuthorization' => $challenge->getPayload(),
+            'token'            => $challenge->getToken(),
+        ];
+
+        $response = $this->httpClient->signedRequest('POST', $challenge->getUrl(), $payload);
+
+        // Waiting loop
+        $endTime = time() + $timeout;
+
+        while (time() <= $endTime) {
+            $response = $this->httpClient->signedRequest('GET', $challenge->getLocation());
+
+            if ('pending' !== $response['status']) {
+                break;
+            }
+
+            sleep(1);
+        }
+
+        if ('pending' === $response['status']) {
+            throw new HttpChallengeTimedOutException($response);
+        } elseif ('valid' !== $response['status']) {
+            throw new HttpChallengeFailedException($response);
+        }
+
+        return $response;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function getCertificatesChain()
+    public function requestCertificate($domain, CertificateRequest $csr, $timeout = 180)
     {
-        return $this->certificatesChain;
+        Assert::stringNotEmpty($domain, 'requestCertificate::$domain expected a non-empty string. Got: %s');
+        Assert::integer($timeout, 'requestCertificate::$timeout expected an integer. Got: %s');
+
+        $humanText = ['-----BEGIN CERTIFICATE REQUEST-----', '-----END CERTIFICATE REQUEST-----'];
+
+        $csrContent = $this->csrSigner->signCertificateRequest($csr);
+        $csrContent = trim(str_replace($humanText, '', $csrContent));
+        $csrContent = trim($this->httpClient->getBase64Encoder()->encode(base64_decode($csrContent)));
+
+        $response = $this->requestResource('POST', ResourcesDirectory::NEW_CERTIFICATE, [
+            'resource' => ResourcesDirectory::NEW_CERTIFICATE,
+            'csr'      => $csrContent,
+        ], false);
+
+        // If the CA has not yet issued the certificate, the body of this response will be empty
+        if (strlen(trim($response)) < 10) { // 10 to avoid false results
+            $location = $this->httpClient->getLastLocation();
+
+            // Waiting loop
+            $endTime = time() + $timeout;
+
+            while (time() <= $endTime) {
+                $response = $this->httpClient->unsignedRequest('GET', $location, null, false);
+
+                if (200 === $this->httpClient->getLastCode()) {
+                    break;
+                }
+
+                if (202 !== $this->httpClient->getLastCode()) {
+                    throw new CertificateRequestFailedException($response);
+                }
+
+                sleep(1);
+            }
+
+            if (202 === $this->httpClient->getLastCode()) {
+                throw new CertificateRequestTimedOutException($response);
+            }
+        }
+
+        // Find issuers certificate
+        $certificatesChain = null;
+
+        foreach ($this->httpClient->getLastLinks() as $link) {
```

Comment:
That's wanted. The request on line 193 can either return directly the certificate (in which case the headers will be present) or an empty body indicating the certificate is not yet ready. If the certificate is not ready, we request it (line 206) until it is, and then we use the last response links.